### PR TITLE
Filter SNMP traps in zentrap

### PR DESCRIPTION
**Pull request by tgjonestx**
A variation of the changes committed to a different release as part of ZEN-10260.

For default filter definition, see zentrap.filter.conf in this services [PR](https://github.com/zenoss/zenoss-service/pull/89)

### CODE REVIEW

File: Products/ZenEvents/TrapFilter.py [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+        @rtype: string
+        """
+        if len(remainingTokens) > 2:
+            return "Too many fields found; at most 4 fields allowed for V1 filters"
+
+        oidOrGenericTrap = ""
+        if len(remainingTokens) > 0:
+            oidOrGenericTrap = remainingTokens[0].strip(".")
+        if len(oidOrGenericTrap) == 1 and oidOrGenericTrap != "*":
+            if not oidOrGenericTrap.isdigit() or not oidOrGenericTrap in "012345":
```
Comment (jplouis) [inline Products/ZenEvents/TrapFilter.py]: oidOrGenericTrap not in "012345"

File: Products/ZenEvents/zentrap.py [hunk 2]
```diff
@@ -760,6 +774,12 @@ class TrapDaemon(CollectorDaemon):
 
     _frameworkFactoryName = "nosip"
 
+    def __init__(self, *args, **kwargs):
+        self._trapFilter = TrapFilter()
+        zope.component.provideUtility(self._trapFilter, ICollectorEventTransformer)
+        super(TrapDaemon, self).__init__(*args, **kwargs)
+        self._trapFilter.initialize()
```
Comment (jplouis) [inline Products/ZenEvents/zentrap.py]: use an initializecallback in the super constructor to init filters.  By the time the init callback is triggered you can send an event to zenoss if the filters have an error.  Be sure to send a clear event if filters are successful or no filters defined.

File: Products/ZenEvents/TrapFilter.py [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+                        # skip blank lines
+                        line = line.strip()
+                        if not line:
+                            continue;
+
+                        errorMessage = self._parseFilterDefinition(line, lineNumber)
+                        if errorMessage:
+                            log.error("Failed to parse filter definition file %s at line %d: %s", format(path), lineNumber, errorMessage)
+                            log.error("Exiting due to invalid filter definition file")
+                            sys.exit(1)
```
Comment (jplouis) [inline Products/ZenEvents/TrapFilter.py]: you can call stop in the init call back and set an exit code there.  Might throw an exception with message here.

File: Products/ZenEvents/TrapFilter.py [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+
+                self._filtersDefined = 0 != (len(self._v1Traps) + len(self._v1Filters) + len(self._v2Filters))
+                if self._filtersDefined:
+                    log.info("Finished reading filter definition file %s", format(path))
+                else:
+                    log.warn("No zentrap filters found in %s", format(path))
+            else:
+                log.error("Could find filter definition file %s", format(path))
+                log.error("Exiting due to invalid filter definition file")
+                sys.exit(1)
```
Comment (jplouis) [inline Products/ZenEvents/TrapFilter.py]: see previous comment

File: Products/ZenEvents/TrapFilter.py (hunk 1 above)
Comment (tgjonestx) [inline Products/ZenEvents/TrapFilter.py]: done

File: Products/ZenEvents/zentrap.py (hunk 2 above)
Comment (tgjonestx) [inline Products/ZenEvents/zentrap.py]: The implementation in 36d5627 only half works - no error event is displayed in RM. Not sure what I'm missing

**Pull request merged**